app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+        if (!ref_smart_interrupt->is_active) {
+            ref_smart_interrupt->position = position;
+            ref_smart_interrupt->config = config;
+            ref_smart_interrupt->is_active = true;
+            ref_smart_interrupt->is_pressed = false;
+            ref_smart_interrupt->first_press = true;
+            *smart_interrupt = ref_smart_interrupt;
+            return 0;
+        }
+    }
+    return -ENOMEM;
+}
+
+static bool is_other_key_shared(struct active_smart_interrupt *smart_interrupt, int32_t position) {
+    for (int i = 0; i < smart_interrupt->config->shared_key_positions_len; i++) {
+        if (smart_interrupt->config->shared_key_positions[i] == position) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool is_layer_shared(struct active_smart_interrupt *smart_interrupt, int32_t layer) {
+    for (int i = 0; i < smart_interrupt->config->shared_layers_len; i++) {
+        if (smart_interrupt->config->shared_layers[i] == layer) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static int on_smart_interrupt_binding_pressed(struct zmk_behavior_binding *binding,
+                                              struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_smart_interrupt_config *cfg = dev->config;
+    struct active_smart_interrupt *smart_interrupt;
+    smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        if (new_smart_interrupt(event.position, cfg, &smart_interrupt) == -ENOMEM) {
+            LOG_ERR("Unable to create new smart_interrupt. Insufficient space in "
+                    "active_smart_interrupts[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        LOG_DBG("%d created new smart_interrupt", event.position);
+    }
+    LOG_DBG("%d smart_interrupt pressed", event.position);
+    stop_timer(smart_interrupt);
+    smart_interrupt->is_pressed = true;
+    if (smart_interrupt->first_press) {
+        behavior_keymap_binding_pressed(&cfg->behaviors[0], event);
+        behavior_keymap_binding_released(&cfg->behaviors[0], event);
+        smart_interrupt->first_press = false;
+    }
+    behavior_keymap_binding_pressed(&cfg->behaviors[1], event);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int on_smart_interrupt_binding_released(struct zmk_behavior_binding *binding,
+                                               struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_smart_interrupt_config *cfg = dev->config;
+    LOG_DBG("%d smart_interrupt keybind released", event.position);
+    struct active_smart_interrupt *smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    smart_interrupt->is_pressed = false;
```

Comment:
Perhaps should be a separate free/release function that captures the intent of this work.